### PR TITLE
feat: add fee percent validations

### PR DESCRIPTION
**Pull request by aleksuss**

### CODE REVIEW

File: fees/src/lib.rs [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
 use std::error::Error;
 use std::fmt::{Display, Formatter};
 use std::num::ParseFloatError;
 use std::str::FromStr;
 
-// We multiply the percent to 100 here to get rid of the floating numbers.
+// We multiply percents to 100 here to get rid of the floating numbers.
+const MIN_FEE_PERCENT: u64 = 1; // 0.01 %
+const MAX_FEE_PERCENT: u64 = 1000; // 10 %
 const DEFAULT_PERCENT: U64 = U64(500); // 5%
```
Comment (raventid) [inline fees/src/lib.rs:12]: Probably we can make MIN/MAX FEE constants U64 type too. This way in the code:

```
    if result < MIN_FEE_PERCENT {
        Err(ParseError::TooLowPercent)
    } else if result > MAX_FEE_PERCENT {
        Err(ParseError::TooHighPercent)
    } else {
        Ok(U64(result))
    }
```

We can wrap result in U64 immediately and use this comparison:

```
    if result < MIN_FEE_PERCENT {
        Err(ParseError::TooLowPercent)
    } else if result > MAX_FEE_PERCENT {
        Err(ParseError::TooHighPercent)
    } else {
        Ok(result)
    }
```

Comment (raventid) [reply fees/src/lib.rs:12]: But in this case U64 type would require to support Eq, PartialEq 🤷

**Review by birchmd — APPROVED** [review]

File: forwarder/src/lib.rs [hunk 2]
```diff
@@ -127,3 +145,31 @@ pub trait ExtFeesCalculator {
         target_address: &str,
     ) -> U128;
 }
+
+// Validate that calculated part of the fee isn't more than `MAX_FEE_PERCENT`.
+fn is_fee_allowed(amount: U128, fee: U128) -> bool {
```
Comment (raventid) [inline forwarder/src/lib.rs:150]: This code is a bit difficult to read (IMHO), maybe plain

```
let percent = fee.0 * 100 / amount.0;
let remainder = (fee.0 * 100) % amount.0;

let allowed_percent = percent < MAX_FEE_PERCENT;
let no_remainder = percent == MAX_FEE_PERCENT && remainder == 0;

allowed_percent && no_remainer
```

**Review by raventid — APPROVED** [review]

File: forwarder/src/lib.rs (hunk 2 above)
Comment (raventid) [reply forwarder/src/lib.rs:150]: Ah, my bad, looking at this second time it's not gonna work this way!

File: fees/src/lib.rs (hunk 1 above)
Comment (aleksuss) [reply fees/src/lib.rs:12]: Yes. That's why I don't wrap the `MIN_FEE_PERCENT` and `MAX_FEE_PERCENT` in the `U64`. And implementation of the traits `Eq` and `PartialEq` on our side is overkill 😄

Comment (raventid) [reply fees/src/lib.rs:12]: Yeah, it works for me this way

File: forwarder/src/lib.rs (hunk 2 above)
Comment (raventid) [reply forwarder/src/lib.rs:150]: Yeah, probably match is OK, but this `zip` looks a bit unnecessary

Comment (aleksuss) [reply forwarder/src/lib.rs:150]: Just for the creation of a combined variable. Provide your variant please?

**Review by mrLSD — APPROVED** [review]

File: forwarder/src/lib.rs (hunk 2 above)
Comment (raventid) [reply forwarder/src/lib.rs:150]: Maybe just a plain wrap in `match()` constructor

```
    let fee = fee.0 * 100;
    let amount = amount.0;

    match (fee.checked_div(amount), fee.checked_rem(amount)) {
        (Some(percent), _) if percent > 100 => false,
        (Some(percent), Some(reminder)) if percent == 100 && reminder > 0 => false,
        _ => true,
    };
```

Comment (aleksuss) [reply forwarder/src/lib.rs:150]: Looks good to me.

**Pull request merged**